projects/components/src/table/data/table-cdk-data-source.ts

```diff
@@ -297,27 +297,24 @@ export class TableCdkDataSource implements DataSource<TableRow> {
 
   private hasCacheForRequest(request: TableDataRequest): boolean {
     if (
-      this.cachedData.rows.length !== 0 &&
-      this.cachedData.rows.length === this.cachedData.total &&
-      request.position.limit <= this.cachedData.rows.length &&
-      !this.haveColumConfigsChanged(request)
+      !isEqual(this.cachedData.request?.sort, request.sort) ||
+      !isEqual(this.cachedData.request?.filters, request.filters) ||
+      this.haveColumConfigsChanged(request)
     ) {
-      // Check if we already have all available results cached
-      return true;
+      // If the columns, sort or filters have changed, we need to make a new request
+      return false;
     }
 
-    if (this.cachedData.rows.length < request.position.limit || !isEqual(this.cachedData.request?.sort, request.sort)) {
```

Comment:
The first clause here is redundant with the final calculation I believe. The second one moved to the top.